examples/vision/deeplabv3_plus.py

```diff
@@ -0,0 +1,298 @@
+
+
+dataset = data_generator(train_images, train_masks)
+dataset
+
+"""
+## Building DeepLabV3+ Model
+"""
+
+"""
+DeepLabv3+ extends DeepLabv3 by employing a encoder-decoder structure. The encoder module
+encodes multi-scale contextual information by applying atrous convolution at multiple
+scales, while the simple yet effective decoder module refines the segmentation results
+along object boundaries.
+
+![](https://github.com/lattice-ai/DeepLabV3-Plus/raw/master/assets/deeplabv3_plus_diagram.png)
+![](https://github.com/lattice-ai/DeepLabV3-Plus/raw/master/assets/deeplabv3_plus_diagram.png)
+
+
+"""
+
+"""
+**Atrous Convolution:** With Atrous convolution, as we go deeper, we can keep the stride
+constant but with larger field-of-view without increasing the number of parameters or the
+amount of computation. And finally, we can have larger output feature map which is good
+for semantic segmentation.
+
+The reason for using **Atrous Spatial Pyramid Pooling** is that it is discovered as the
+sampling rate becomes larger, the number of valid filter weights (i.e., the weights that
+are applied to the valid feature region, instead of padded zeros) becomes smaller.
+"""
+
+
+def convolution_block(
+    block_input,
+    n_filters=256,
+    kernel_size=3,
+    dilation_rate=1,
+    padding="same",
+    use_bias=False,
+):
+
+    x = tf.keras.layers.Conv2D(
+        n_filters,
+        kernel_size=1,
+        dilation_rate=1,
+        padding="same",
+        use_bias=use_bias,
+        kernel_initializer=tf.keras.initializers.HeNormal(),
+    )(block_input)
+    x = tf.keras.layers.BatchNormalization()(x)
+    return tf.nn.relu(x)
+
+
+def AtrousSpatialPyramidPooling(aspp_input):
+
+    dims = tf.keras.backend.int_shape(aspp_input)
+
+    layer = tf.keras.layers.AveragePooling2D(pool_size=(dims[-3], dims[-2]))(aspp_input)
+    layer = convolution_block(layer, kernel_size=1, use_bias=True)
+
+    out_pool = tf.keras.layers.UpSampling2D(
+        size=(dims[-3] // layer.shape[1], dims[-2] // layer.shape[2]),
+        interpolation="bilinear",
+    )(layer)
+
+    out_1 = convolution_block(aspp_input, kernel_size=1, dilation_rate=1)
+    out_6 = convolution_block(aspp_input, kernel_size=3, dilation_rate=6)
+    out_12 = convolution_block(aspp_input, kernel_size=3, dilation_rate=12)
+    out_18 = convolution_block(aspp_input, kernel_size=3, dilation_rate=18)
+
+    layer = tf.keras.layers.Concatenate(axis=-1)(
+        [out_pool, out_1, out_6, out_12, out_18]
+    )
+
+    output = convolution_block(layer, kernel_size=1)
+
+    return output
+
+
+"""
+The encoder features are first bilinearly upsampled by a factor of 4 and then 
+concatenated with the corresponding low-level features from the network backbone that
+have the same spatial resolution. For this example, we would be 
+using Resnet101 pre-trained on ImageNet as the backbone model and we would use 
+the low-level features from the Conv2 block of the backbone.
+"""
+
+
+def DeeplabV3Plus():
```

Comment:
You probably want to make `image_size` and `num_classes` signature arguments

Reply:
Passed `image_size` and `num_classes` as arguments